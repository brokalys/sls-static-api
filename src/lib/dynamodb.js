import dynamodb from 'serverless-dynamodb-client';

export async function batchGet(table, keys) {
  const { Responses } = await dynamodb.doc
    .batchGet({
      RequestItems: {
        [table]: {
          Keys: keys,
        },
      },
    })
    .promise();

  return Responses[table];
}

export function put(table, data) {
  return dynamodb.doc
    .put({
      TableName: table,
      Item: data,
    })
    .promise();
}

export default { batchGet, put };
