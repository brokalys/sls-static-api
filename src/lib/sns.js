import AWS from 'aws-sdk';

import isOffline from './offline';

const sns = new AWS.SNS({
  apiVersion: '2010-03-31',
  endpoint: isOffline() ? 'http://127.0.0.1:4002' : undefined,
});

const accountId = isOffline() ? 123456789012 : 173751334418;

export function publish(message) {
  return sns
    .publish({
      Message: JSON.stringify({
        default: JSON.stringify(message),
      }),
      MessageStructure: 'json',
      TopicArn: `arn:aws:sns:${process.env.AWS_REGION}:${accountId}:${process.env.APP_STAGE}-load-properties`,
    })
    .promise();
}

export default { publish };
