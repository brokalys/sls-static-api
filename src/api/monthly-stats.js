import { latviaRelationships } from '@brokalys/location-json-schemas';
import Joi from 'joi';
import numbers from 'numbers';

import dynamodb from '../lib/dynamodb';
import createHash from '../lib/hash';
import moment from '../lib/moment';
import sqs from '../lib/sqs';

import Bugsnag from '@bugsnag/js';
import BugsnagPluginAwsLambda from '@bugsnag/plugin-aws-lambda';

const busgnagLogger = {
  debug: function () {},
  info: console.info,
  warn: console.warn,
  error: console.error,
};

Bugsnag.start({
  apiKey: process.env.BUGSNAG_KEY,
  logger: busgnagLogger,
  releaseStage: process.env.STAGE,
  enabledReleaseStages: ['staging', 'prod'],
  plugins: [BugsnagPluginAwsLambda],
});

const VALID_LOCATION_IDS = Array.from(
  Object.entries(latviaRelationships).reduce(
    (carry, [locationId, children]) => {
      carry.add(locationId);
      children.forEach((child) => carry.add(child));
      return carry;
    },
    new Set(),
  ),
);

const validationSchema = Joi.object({
  discard: Joi.number().min(0).max(1).precision(2),
  start_datetime: Joi.date()
    .default('2018-01-01')
    .when(Joi.ref('/source'), {
      is: 'real-sales',
      then: Joi.date().min('1998-01-01'),
      otherwise: Joi.date().min('2018-01-01'),
    }),
  end_datetime: Joi.date()
    .default(moment.utc().format('YYYY-MM-DD'))
    .min(Joi.ref('start_datetime'))
    .max(new Date()),
  filters: Joi.object({
    category: Joi.string().valid('apartment', 'house', 'land').required(),
    type: Joi.string().valid('sell', 'rent').when(Joi.ref('/source'), {
      is: 'real-sales',
      otherwise: Joi.required(),
    }),
    location_classificator: Joi.string()
      .valid(...VALID_LOCATION_IDS)
      .when(Joi.ref('/source'), {
        is: 'real-sales',
        then: Joi.required(),
      }),
  }).required(),
  source: Joi.string()
    .valid('classifieds', 'real-sales')
    .default('classifieds'),
  period: Joi.string().valid('month', 'quarter', 'year').default('month'),
});

function decodeQuerystring(qs) {
  function decodeValue(value) {
    try {
      return JSON.parse(decodeURIComponent(value));
    } catch (e) {
      return value;
    }
  }
  return Object.entries(qs || {}).reduce(
    (carry, [key, value]) => ({
      ...carry,
      [key]: decodeValue(value),
    }),
    {},
  );
}

function prepareSearchQueries(filters, dates, source) {
  return dates
    .map(([startDate, endDate]) => ({
      filters: {
        ...filters,
        type: source === 'classifieds' ? filters.type : 'sell',
      },
      start_datetime: startDate.toISOString(),
      end_datetime: endDate.toISOString(),
      source: source === 'classifieds' ? undefined : source,
    }))
    .map((self) => ({
      ...self,
      hash: createHash(self),
    }));
}

function calculateStatistics(prices) {
  return {
    min: Math.ceil(numbers.basic.min(prices)) || null,
    max: Math.ceil(numbers.basic.max(prices)) || null,

    mean: Math.ceil(numbers.statistic.mean(prices)) || null,
    mode: Math.ceil(numbers.statistic.mode(prices)) || null,
    median: Math.ceil(numbers.statistic.median(prices)) || null,
  };
}

function discardData(data, amount = 0) {
  const itemCount = data.length;
  const discardStart = Math.floor((itemCount * amount) / 2);

  return data
    .sort((a, b) => a - b)
    .splice(discardStart, itemCount - discardStart * 2);
}

export const handler = async (event) => {
  const queryStringParameters = decodeQuerystring(event.queryStringParameters);
  const validation = validationSchema.validate(queryStringParameters);

  if (validation.error) {
    return {
      statusCode: 400,
      headers: {
        'Access-Control-Allow-Origin': '*',
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        errors: validation.error.details,
      }),
    };
  }

  const { start_datetime, end_datetime, filters, discard, source, period } =
    validation.value;

  const startDate = moment.utc(start_datetime);
  const endDate = moment.utc(end_datetime);

  const momentRange = moment().range(startDate, endDate);
  const dates = Array.from(momentRange.by(period, { excludeEnd: true })).map(
    (date) => [date, date.clone().endOf(period)],
  );

  // Build search queries for all dates
  const searchQueries = prepareSearchQueries(filters, dates, source);

  // Load the dynamodb results
  const data = await dynamodb.batchGet(
    process.env.CACHE_DYNAMODB_TABLE_NAME,
    searchQueries.map(({ hash }) => ({ hash })),
  );

  // Calculate how many more results are missing
  const loadingResults = searchQueries.length - data.length;

  // Trigger SQS actions for each result that's not found in the archive
  if (loadingResults > 0) {
    const foundRows = data.map(({ hash }) => hash);
    const notFoundRows = searchQueries.filter(
      ({ hash }) => !foundRows.includes(hash),
    );

    await Promise.all(notFoundRows.map((row) => sqs.sendMessage(row)));
  }

  const results = data
    .map((row) => {
      const prices = discardData(row.prices, discard);
      const pricesPerSqm = discardData(row.pricesPerSqm, discard);

      return {
        count: prices.length,
        start_datetime: row.start_datetime,
        end_datetime: row.end_datetime,

        price: calculateStatistics(prices),
        pricePerSqm: calculateStatistics(pricesPerSqm),
      };
    })
    .sort((a, b) =>
      moment(a.start_datetime).isAfter(b.start_datetime) ? 1 : -1,
    );

  return {
    statusCode: 200,
    headers: {
      'Access-Control-Allow-Origin': '*',
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      loadingResults,
      results,
    }),
  };
};

const bugsnagHandler = Bugsnag.getPlugin('awsLambda').createHandler();
export const run = bugsnagHandler(handler);
