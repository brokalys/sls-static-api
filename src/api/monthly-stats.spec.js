import { run } from './monthly-stats';

import dynamodb from '../lib/dynamodb';
import sns from '../lib/sns';

jest.mock('../lib/dynamodb');
jest.mock('../lib/sns');

describe('monthly-stats', () => {
  beforeEach(jest.resetAllMocks);

  test.each([
    encodeURIComponent(JSON.stringify({ category: 'APARTMENT', type: 'sell' })), // case sensitive
    encodeURIComponent(JSON.stringify({ category: 'apartment', type: 'SELL' })), // case sensitive
    encodeURIComponent(JSON.stringify({ category: 'unknown', type: 'sell' })), // unknown values
    encodeURIComponent(
      JSON.stringify({ category: 'apartment', type: 'unknown' }),
    ), // unknown values
    encodeURIComponent(JSON.stringify({ unknown: 'apartment', type: 'sell' })), // unknown keys
    encodeURIComponent(JSON.stringify({})), // missing category
    encodeURIComponent({ category: 'apartment', type: 'sell' }), // not JSON-encoded

    // wrong datatypes
    '',
    'wrong',
    undefined,
    null,
    123,
    1.23,
    true,
    false,
    {},
  ])(
    'returns a validation error if filter validation fails with: %# - %j',
    async (input) => {
      const output = await run({
        queryStringParameters: {
          filters: input,
        },
      });

      expect(output).toEqual(
        expect.objectContaining({
          statusCode: 400,
        }),
      );
    },
  );

  test.each([
    { filters: {} }, // missing filters
    { filters: { category: 'apartment' } }, // missing type filter
    { filters: { type: 'sell' } }, // missing category filter
    {
      filters: { category: 'apartment', type: 'sell' },
      end_datetime: '2050-01-01',
    }, // cannot be in future
    {
      filters: { category: 'apartment', type: 'sell' },
      start_datetime: '2020-02-01',
      end_datetime: '2020-01-01',
    }, // start cannot be before end
    {
      filters: { category: 'apartment', type: 'sell' },
      start_datetime: '2017-12-01',
    }, // start cannot be before 2018-01-01
    { discard: 1.1 }, // too large discard
    { discard: -1 }, // too small discard
    { discard: 0.133 }, // precision wrong (max: 2)

    // wrong datatypes
    '',
    'wrong',
    undefined,
    null,
    123,
    1.23,
    true,
    false,
    {},
  ])(
    'returns a validation error if root-level validation fails with: %# - %j',
    async (input) => {
      const output = await run({
        queryStringParameters: input,
      });

      expect(output).toEqual(
        expect.objectContaining({
          statusCode: 400,
        }),
      );
    },
  );

  test('returns the data upon success', async () => {
    dynamodb.batchGet.mockReturnValueOnce([
      {
        hash: 'third_hash',
        start_datetime: '2018-03-01T00:00:00.000Z',
        prices: [],
        pricesPerSqm: [],
      },
      {
        hash: 'first_hash',
        start_datetime: '2018-01-01T00:00:00.000Z',
        prices: [100],
        pricesPerSqm: [10],
      },
      {
        hash: 'second_hash',
        start_datetime: '2018-02-01T00:00:00.000Z',
        prices: [200, 300],
        pricesPerSqm: [20, 30],
      },
    ]);

    const output = await run({
      queryStringParameters: {
        start_datetime: '2018-01-01',
        end_datetime: '2018-04-01',
        filters: { category: 'apartment', type: 'sell' },
      },
    });

    expect(output).toMatchSnapshot();
    expect(sns.publish).not.toBeCalled();
  });

  test.only('discards 50% of the results', async () => {
    dynamodb.batchGet.mockReturnValueOnce([
      {
        hash: 'third_hash',
        start_datetime: '2018-03-01T00:00:00.000Z',
        prices: [100, 200, 300],
        pricesPerSqm: [10, 20, 30],
      },
      {
        hash: 'first_hash',
        start_datetime: '2018-01-01T00:00:00.000Z',
        prices: [100],
        pricesPerSqm: [10],
      },
      {
        hash: 'second_hash',
        start_datetime: '2018-02-01T00:00:00.000Z',
        prices: [200, 300, 400, 500],
        pricesPerSqm: [20, 30, 40, 50],
      },
    ]);

    const output = await run({
      queryStringParameters: {
        discard: 0.5,
        start_datetime: '2018-01-01',
        end_datetime: '2018-04-01',
        filters: { category: 'apartment', type: 'sell' },
      },
    });

    expect(output).toMatchSnapshot();
    expect(sns.publish).not.toBeCalled();
  });

  test('triggers a SNS to load the complete dataset if something is missing', async () => {
    dynamodb.batchGet.mockReturnValueOnce([
      {
        hash: 'first_hash',
        prices: [100],
        pricesPerSqm: [10],
      },
    ]);

    await run({
      queryStringParameters: {
        start_datetime: '2018-01-01',
        end_datetime: '2018-03-01',
        filters: { category: 'apartment', type: 'sell' },
      },
    });

    expect(sns.publish).toBeCalled();
    expect(sns.publish).not.toBeCalledWith({
      hash: 'first_hash',
      filters: {
        category: 'apartment',
      },
    });
  });
});
