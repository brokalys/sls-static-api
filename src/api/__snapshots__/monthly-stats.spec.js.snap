// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`monthly-stats discards 50% of the results 1`] = `
Object {
  "body": "{\\"loadingResults\\":0,\\"results\\":[{\\"count\\":1,\\"start_datetime\\":\\"2018-01-01T00:00:00.000Z\\",\\"price\\":{\\"min\\":100,\\"max\\":100,\\"mean\\":100,\\"mode\\":100,\\"median\\":100},\\"pricePerSqm\\":{\\"min\\":10,\\"max\\":10,\\"mean\\":10,\\"mode\\":10,\\"median\\":10}},{\\"count\\":2,\\"start_datetime\\":\\"2018-02-01T00:00:00.000Z\\",\\"price\\":{\\"min\\":300,\\"max\\":400,\\"mean\\":350,\\"mode\\":300,\\"median\\":350},\\"pricePerSqm\\":{\\"min\\":30,\\"max\\":40,\\"mean\\":35,\\"mode\\":30,\\"median\\":35}},{\\"count\\":3,\\"start_datetime\\":\\"2018-03-01T00:00:00.000Z\\",\\"price\\":{\\"min\\":100,\\"max\\":300,\\"mean\\":200,\\"mode\\":100,\\"median\\":200},\\"pricePerSqm\\":{\\"min\\":10,\\"max\\":30,\\"mean\\":20,\\"mode\\":10,\\"median\\":20}}]}",
  "headers": Object {
    "Content-Type": "application/json",
  },
  "statusCode": 200,
}
`;

exports[`monthly-stats returns the data upon success 1`] = `
Object {
  "body": "{\\"loadingResults\\":0,\\"results\\":[{\\"count\\":1,\\"start_datetime\\":\\"2018-01-01T00:00:00.000Z\\",\\"price\\":{\\"min\\":100,\\"max\\":100,\\"mean\\":100,\\"mode\\":100,\\"median\\":100},\\"pricePerSqm\\":{\\"min\\":10,\\"max\\":10,\\"mean\\":10,\\"mode\\":10,\\"median\\":10}},{\\"count\\":2,\\"start_datetime\\":\\"2018-02-01T00:00:00.000Z\\",\\"price\\":{\\"min\\":200,\\"max\\":300,\\"mean\\":250,\\"mode\\":200,\\"median\\":250},\\"pricePerSqm\\":{\\"min\\":20,\\"max\\":30,\\"mean\\":25,\\"mode\\":20,\\"median\\":25}},{\\"count\\":0,\\"start_datetime\\":\\"2018-03-01T00:00:00.000Z\\",\\"price\\":{\\"min\\":null,\\"max\\":null,\\"mean\\":null,\\"mode\\":null,\\"median\\":null},\\"pricePerSqm\\":{\\"min\\":null,\\"max\\":null,\\"mean\\":null,\\"mode\\":null,\\"median\\":null}}]}",
  "headers": Object {
    "Content-Type": "application/json",
  },
  "statusCode": 200,
}
`;
